src/opencl_sha256crypt_fmt_plug.c

```diff
@@ -545,39 +564,47 @@ static int crypt_all(int *pcount, struct db_salt *_salt)
 {
 	int count = *pcount;
 	int index;
-	int *indices = NULL;  // relationship between sorted and unsorted plaintext list
 	size_t gws;
 	size_t *lws = local_work_size ? &local_work_size : NULL;
 	sha256_password *input_candidates;
 	sha256_hash *output_hashes;
 
 	gws = GET_NEXT_MULTIPLE(count, local_work_size);
 
-	input_candidates = plaintext;
-	output_hashes = calculated_hash;
+	if (bitmap_of_lens & (bitmap_of_lens - 1)) {
+		input_candidates = plain_sorted;
+		output_hashes = computed_hash;
+	} else {
+		input_candidates = plaintext;
+		output_hashes = calculated_hash;
+	}
 
 	if (new_keys) {
 		// sort passwords by length
 		int tot_todo = 0, len;
 
-		if (lens_differ) {
-			indices = mem_alloc(gws * sizeof(int));
+		if (bitmap_of_lens & (bitmap_of_lens - 1)) {
+			if (gws > indices_size) {
+				indices_size = gws;
+				MEM_FREE(indices);
+				indices = mem_alloc(gws * sizeof(int));
+			}
 
 			for (len = 0; len <= PLAINTEXT_LENGTH; len++) {
-				for (index = 0; index < count; index++) {
-					if (plaintext[index].length == len)
-						indices[tot_todo++] = index;
-				}
+				if (!(bitmap_of_lens >> len))
+					break;
+				if (bitmap_of_lens & (1 << len))
```

Comment:
Please write this check exactly as I had suggested. Your way of writing it kills the common subexpression, which the compiler would optimize (compute it just once).